### ISSUE TITLE
conversationUpdate not emitted when Bot joins a Teams channel

### ISSUE THREAD

**Issue by sferdi0** [microsoft/botbuilder-python#1469]
I need to detect if my bot is added to a Teams channel, but I don't receive the relevant `conversationUpdate` event. From documentation we can see [this notification](https://docs.microsoft.com/en-us/microsoftteams/platform/resources/bot-v3/bots-notifications#schema-example-bot-added-to-team), but I'm not receiving it.

### Reproduce

I open Microsoft Teams, side-load my app, click on my app, click `Add to Team`, pick a channel, and I get this event (anonymized):

```text
T 2021/01/18 11:36:39.505883 127.0.0.1:34130 -> 127.0.0.1:8090 [AP] #24
POST /api/messages HTTP/1.0
Host: xxx.dev
X-Real-IP: xxx
X-Forwarded-For: xxx
X-Forwarded-Proto: https
Connection: close
Content-Length: 966
Authorization: Bearer xxxx.
User-Agent: Microsoft-SkypeBotApi (Microsoft-BotFramework/3.0)
MS-CV: xxxx.1.1.1.2773549121.1.2
Content-Type: application/json; charset=utf-8

{
  "membersAdded": [
    {
      "id": "29:1sMLH5PmHN-_fqpvKrbUd-PAbPJltPLkXa4Z6qxxxxx",
      "aadObjectId": "f5fd3ae4-xxxx"
    },
    {
      "id": "28:539767ce-bd98-4bdb-xxxx-xxxxx"
    }
  ],
  "type": "conversationUpdate",
  "timestamp": "2021-01-18T11:36:39.0083926Z",
  "id": "f:9b9dadc6-xxxx-xxxx",
  "channelId": "msteams",
  "serviceUrl": "https://smba.trafficmanager.net/emea/",
  "from": {
    "id": "29:1sMLH5PmHN-_fqpvKrbUd-PAbPJltPLkXa4Z6qASBi9M_NerI_xxxxxxxx",
    "aadObjectId": "f5fd3ae4-1798-4dd4-xxxx-xxxx"
  },
  "conversation": {
    "conversationType": "personal",
    "tenantId": "ed78c1db-7837-4e88-90e8-xxxx",
    "id": "a:1E7Ma2tnUsKRtWES25AK7z1_besIphdwhIFXXXMq_aQZcyYXXXXX"
  },
  "recipient": {
    "id": "28:539767ce-bd98-XXXX",
    "name": "BotTeamsTestXXX"
  },
  "channelData": {
    "tenant": {
      "id": "ed78c1db-7837-4e88-XXXX-XXXX"
    }
  }
}
```

Which is a personal event, it is mostly concerning the user that has added my app to their Team, and I can use this to private message them (welcome message).

However, from [the documentation](https://docs.microsoft.com/en-us/microsoftteams/platform/resources/bot-v3/bots-notifications#bot-or-user-added-to-a-team) I can read:

> The `conversationUpdate` event with the `membersAdded` object in the payload is sent **when either a bot is added to a team** or a new user is added to a team where a bot has been added.
> Because this event is sent in both cases, you should parse the `membersAdded` object to determine ....

The documentation also [gives us the payload](https://docs.microsoft.com/en-us/microsoftteams/platform/resources/bot-v3/bots-notifications#schema-example-bot-added-to-team) of when it joins a Teams channel:

```json
   "channelData":{
      "team":{
         "id":"19:efa9296d959346209fea44151c742e73@thread.skype"
      },
      "eventType":"teamMemberAdded",
      "tenant":{
         "id":"72f988bf-86f1-41af-91ab-2d7cd011db47"
      }
   }
```

From the network traffic that I posted, I can see that I have a `channelData` object, however, I do not have `channelData.team.id`.

In short, I have no idea when my bot is added to a Teams channel. I can only establish that an individual user has added my bot to a Team. I never receive the channel event. I think this a bug, so I'm posting it here.

**Comment by sferdi0** [comment]
For reference, here is my app manifest, anonymized;

```json
{
    "$schema": "https://raw.githubusercontent.com/OfficeDev/microsoft-teams-app-schema/preview/DevPreview/MicrosoftTeams.schema.json",
    "accentColor": "#FFFFFF",
    "bots": [
        {
            "botId": "539767ce-bd98-4bdb-xxxx-xxxx",
            "isNotificationOnly": false,
            "needsChannelSelector": false,
            "scopes": [
                "personal",
                "team",
                "groupchat"
            ],
            "supportsFiles": false
        }
    ],
    "configurableTabs": [
        {
            "canUpdateConfiguration": true,
            "configurationUrl": "https://xxx.dev/configure",
            "context": [
                "channelTab",
                "privateChatTab",
                "meetingStage",
                "meetingChatTab",
                "meetingDetailsTab",
                "meetingSidePanel"
            ],
            "scopes": [
                "team",
                "groupchat"
            ]
        }
    ],
    "description": {
        "full": "A powerful Teams app.",
        "short": "Test!"
    },
    "developer": {
        "name": "Test",
        "privacyUrl": "https://xxx.dev/privacy",
        "termsOfUseUrl": "https://xxx.dev/tos",
        "websiteUrl": "https://xxx.dev"
    },
    "devicePermissions": [
        "media",
        "notifications",
        "openExternal"
    ],
    "icons": {
        "color": "xxx_192x192.png",
        "outline": "xxx_32x32.png"
    },
    "id": "539767ce-bd98-4bdb-xxxx-xxxx",
    "manifestVersion": "devPreview",
    "name": {
        "full": "xxx",
        "short": "xxx"
    },
    "packageName": "com.microsoft.teams.devapp",
    "permissions": [
        "identity",
        "messageTeamMembers"
    ],
    "staticTabs": [
        {
            "contentUrl": "https://xxx.dev/content/inteams",
            "entityId": "personalTabID",
            "name": "Personal tab",
            "scopes": [
                "personal"
            ],
            "websiteUrl": "https://xxx.dev/content/inbrowser"
        },
        {
            "contentUrl": "https://xxx.dev/tab",
            "entityId": "proactiveTabEntityId",
            "name": "Proative Messaging Sample Tab",
            "scopes": [
                "personal"
            ]
        }
    ],
    "validDomains": [
        "xxx.dev"
    ],
    "version": "1.0.0",
    "webApplicationInfo": {
        "id": "539767ce-bd98-4bdb-xxxx-xxxx",
        "resource": "api://xxx.dev/539767ce-bd98-4bdb-xxxx-xxxx"
    }
}
```

**Comment by rvinothrajendran** [comment]
@sferdi0 have you tried  on_teams_members_added event ?

**Comment by sferdi0** [comment]
@rvinothrajendran I have something like

```python3
class BotHandler(TeamsActivityHandler):
    async def on_teams_members_added(
        self,
        teams_members_added: [TeamsChannelAccount],
        team_info: TeamInfo,
        turn_context: TurnContext,
    ):
        teams_members_added = [
            ChannelAccount().deserialize(member.serialize())
            for member in teams_members_added
        ]

        for member_added in teams_members_added:
            if member_added.id != turn_context.activity.recipient.id:
                await turn_context.send_activity("Hello and welcome!")
```

Which sends a message to the user who installed my app.

I did not include it in my bug report because I can tell by observing the network traffic that I'm not getting any event related to my bot joining a channel.

**Comment by rvinothrajendran** [comment]
@sferdi0 Have you received the event if add any other member ?

**Comment by sferdi0** [comment]
Extra information:

- Note that my app manifest is using the dev schema
- I am running the Teams application on Linux, in dev mode.

**Comment by sferdi0** [comment]
> @sferdi0 Have you received the event if add any other member ?

I have not tried this yet. I am sure it would work... I can try tomorow.

**Comment by sferdi0** [comment]
@rvinothrajendran I added a new Azure AD user, logged into my Team from another computer with this account, and my bot, which is in channel `#general` received the following notification (taken from network capture):

```json
{
  "membersAdded": [
    {
      "id": "29:1EbSpmX4cIYvuVK6ktqTMQn3tX74e2oIEnxJxV2UOXlIX-v-EdAnoh2Cj_XXX",
      "aadObjectId": "54f7c6a0-c1ae-4dcc-XXXX-XXXX"
    }
  ],
  "type": "conversationUpdate",
  "timestamp": "2021-01-19T08:42:33.3436845Z",
  "id": "f:4758e44a-f08e-9362-e635-cec816037350",
  "channelId": "msteams",
  "serviceUrl": "https://smba.trafficmanager.net/emea/",
  "from": {
    "id": "29:1ILKbYl9rxDYYE6ZWsy4SHNWKaTdfAKCVPL_a7b-g8ew2SdusXXXX",
    "aadObjectId": "f5fd3ae4-1798-4dd4-XXXX-XXXX"
  },
  "conversation": {
    "isGroup": true,
    "conversationType": "channel",
    "tenantId": "ed78c1db-7837-4e88-XXXX-XXXX",
    "id": "19:95b25f4767234d29a8XXXX@thread.tacv2"
  },
  "recipient": {
    "id": "28:1660bdf0-bdc8-46e0-XXXX-XXXX",
    "name": "XXXX"
  },
  "channelData": {
    "team": {
      "aadGroupId": "6a3f5871-3ba8-4fdf-XXXX-XXXX",
      "name": "My teams' name",
      "id": "19:95b25f4767234d29a83XXXX@thread.tacv2"
    },
    "eventType": "teamMemberAdded",
    "tenant": {
      "id": "ed78c1db-7837-4e88-XXXX-XXXX"
    }
  }
}
```

So inside my Python code I was able to catch this with `on_teams_members_added()`, the `TeamsInfo` was my Team + thread_id, and it had a `ChannelAccount` object of the added member. 

In short, adding (new) members to my Teams channel in which my Bot is in will yield a `conversationUpdate`.

However, when my Bot is added a Teams channel I don't get the `teamMemberAdded` event type, I don't get any event other than the one in my original post.

**Comment by rvinothrajendran** [comment]
@sferdi0  I think Its working for me.  
comment this code  #if member_added.id != turn_context.activity.recipient.id: 

Vinoth is a user 
None is a Bot

![image](https://user-images.githubusercontent.com/16264167/105060428-b12ca300-5a78-11eb-9ac8-c1397e68a478.png)

**Comment by sferdi0** [comment]
@rvinothrajendran what is working for you? Please be more specific.

**Comment by rvinothrajendran** [comment]
Sorry misunderstanding 😒

**Comment by sferdi0** [comment]
Unfortunately this issue is quite a showstopper for me. If I can't detect when my bot has entered a Teams channel, I can't post a welcome message. I can't (proactively) post to a Teams channel at all. 

To re-iterate, the documentation states that when my Bot enters a Teams channel, I am supposed to receive the following message:

- a `conversationUpdate`
  - `conversationType` set to `channel`
  - a `channelData` object
    - containing a `team.id` to get the ID of the Teams channel
    - `eventType` of `teamMemberAdded`
    - A `tenant` id

I have created a "Bot Channels Registration" - and it is "bound" to my App ID within Azure AD. The bot works. Except that I'm not receiving this event.

**Comment by sferdi0** [comment]
Another thing to note is that https://dev.botframework.com/bots says:

> You don't have any bots or skills yet.
> Build a bot using the Bot Builder SDK and optionally provision your bot in Microsoft Azure.
> Create a bot with the Azure Bot Service

Whilst inside Azure (https://portal.azure.com/#home) I have a "Bot Channels Registration" with my bot. Confusing.

**Comment by rvinothrajendran** [comment]
@sferdi0 we will speak in skype now ?

**Comment by sferdi0** [comment]
@rvinothrajendran yes please :-) `sander@sanderf.nl`

**Comment by sferdi0** [comment]
@EricDahlvang

I'm starting to believe that what I want to do is actually not possible, I'm pinging you for a sanity check :-)

In short;

1. My bot enters a Teams channel (for the first time)
2. My bot send a message to this channel "Hello, I am a bot and I have just been added to this channel"

Important: between step 1 and 2 there have been no other activities (such as another member getting added or removed, that my handler can pick up). My bot simply needs to "announce it's presence" immediately upon joining. 

This is a slimmed down version of what my bot needs to do. It needs to be "aware" of what Teams channels exist, so that it can proactively send messages to this channel, at will.

**Comment by sferdi0** [comment]
Well, I'm just talking into a void here... I will come to the conclusion that what I want to do is not possible.

Every other platform I have coded bots for (Slack, IRC, XMPP) will allow you to post a message into a channel. One could even state that a bot's primary goal is to post messages to a channel. Yet, with Microsoft's Bot framework - it is impossible. I will close the issue as I seem to be wanting something *exotic*. Just Microsoft being Microsoft I suppose.

**Comment by EricDahlvang** [comment]
Hi @sferdi0 

I'm sorry you've run into this issue.  I've just created a new teams bot, and added it to a Team.  The initial time the bot is added, it should receive:

1) a conversationUpdate for the bot being added to the user's personal scope
2) a conversationUpdate for the bot being added to the team
3) an installationUpdate for the bot added to the user's personal scope
4) an installationUpdate for the bot added to the team

If the bot is uninstalled, and re-installed, the personal scope conversationUpdates will not arrive again but the team conversationUpdates should.  If you are not seeing those, then there could be a bug.

I assume you've looked over https://github.com/microsoft/BotBuilder-Samples/tree/main/samples/python/58.teams-start-thread-in-channel and https://github.com/microsoft/BotBuilder-Samples/tree/main/samples/python/57.teams-conversation-bot If i'm understanding your requirements, one of these two should provide the functionality you are looking for.

**Comment by sferdi0** [comment]
@EricDahlvang 

The only event I receive is the network traffic I posted in my initial post. This event is emitted by MS when an user installs the app (and places the app into a channel `#general`). From this alone you cannot derive any channels (again, see network traffic), thus you cannot start a thread in a channel, because you don't know about it. This JSON message is obtained from a network capture.

And yes, I have seen both projects. In fact, I've seen most samples. I have code ready to go to start threads in Channel. The problem is that it needs some sort of reference (channel ID). This reference you are supposed to get when an user places the app inside a channel (?). 

If you could kindly test this for me:

1. Create a bot and install it into a Teams channel (`#general`)
2. The bot immediately posts a message "hello" into that channel

**Comment by EricDahlvang** [comment]
Are you uninstalling the bot from the Team before re-installing?

```python
    async def on_installation_update_add(  # pylint: disable=unused-argument
            self, turn_context: TurnContext
        ):
        
        if "settings" in turn_context.activity.channel_data:
            settings = turn_context.activity.channel_data['settings']
            if "selectedChannel" in settings:
                channelId = settings['selectedChannel']['id']
                await turn_context.send_activity(
                        f"Hello, I am a bot and I have just been added to this channel: {channelId}"
                    )
```
![image](https://user-images.githubusercontent.com/11055362/105517058-69ca3080-5c8b-11eb-84dd-ae70e160e01f.png)


Also, technically bots are not added/removed from channels, but are added/removed from a Team which contains channels.  When a bot is installed into a Team, it can retrieve the list of channels for that Team and communicate on them directly via that means.

**Comment by EricDahlvang** [comment]
And don't eat your shoe.  There's a lot to digest here already :)

**Comment by sferdi0** [comment]
@EricDahlvang Thanks for testing. However, I cannot reproduce.

I have created a sample bot here: [sferdi0/teams_bot_bug](https://github.com/sferdi0/teams_bot_bug)

This codebase is based off one of the official samples and should be a minimal working example.

## Installation

```bash
git clone https://github.com/sferdi0/teams_bot_bug.git teams_bot_bug
cd $_
virtualenv -p /usr/bin/python3 venv
pip install -r requirements.txt
MicrosoftAppId="APP_ID_HERE" MicrosoftAppPassword="SECRET" python3 app.py
```

Alternatively before running `app.py` change the `PORT` in `config.py`, as it defaults to `8090`.

### Preparing the app packages

`manifest.json` will need to be modified, change these values:

- `APP_ID_HERE`
- `YOUR_DOMAIN_HERE`

After which we can create a `bla.zip`:

```bash
cd teams_app_manifest
zip bla.zip *   
  adding: icon-color.png (deflated 5%)
  adding: icon-outline.png (stored 0%)
  adding: manifest.json (deflated 58%
```

## Background info

I am running:

- Ubuntu 20.04
- Microsoft Teams Version `1.3.00.30857` (64-bit)
- Dev preview enabled

![](https://i.imgur.com/m2jIyL8.png)

### Azure AD

Inside Azure AD, I have an app registration `PythonTest` which gives me AppID `539767ce-bd98-4bdb-8e72-a1a90a4d1167`. It has an "Application ID URI" of `api://MY_DOMAIN/MY_APP_ID`. 

### Azure Bot Services

I have a "Bot Channels Registration":

- "Messaging endpoint" set to `https://MY_DOMAIN/api/messages`
- "Microsoft App ID" set to `MY_APP_ID`
- It has the "Teams Channel" option enabled.
- Subscription: Pay-As-You-Go
- Bot Service pricing tier: F0

## Testing our app

Make sure `app.py` is running, and inside Teams go into `Apps -> Upload a custom app -> Upload for <tenant_name>`.

![](https://i.imgur.com/WlkzP6o.png)

Upload the recently created `bla.zip`.

![](https://i.imgur.com/PG87LI5.png)

If the upload was successful, it will be listed in the Apps screen:

![](https://i.imgur.com/tV107rq.png)

Click on it and add it to a Team:

![](https://i.imgur.com/pFzxe4f.png)

Add it to `#general`

![](https://i.imgur.com/8HcUjhq.png)

### Result

It results in the bot "joining the channel" (technically it merely joins a Team, that has channels) ... anyway, no channel message is posted:

![](https://i.imgur.com/mnxTtHW.png)

We can however interact with the bot, because we have this in our handler:

```python3
async def on_message_activity(self, turn_context: TurnContext):
    await turn_context.send_activity("Hello :)")
```

![](https://i.imgur.com/OMetppP.png)

Just to show that my `/api/messages` is working and can respond to messages.

### Uninstalling

You asked how I deleted my Bot. 

> Are you uninstalling the bot from the Team before re-installing?

The only way I can remove the bot is by going into `Apps` again and deleting it:

![](https://i.imgur.com/voy0Nhr.png)

![](https://i.imgur.com/5fCXXkv.png)

### Conclusion

I have tried to be specific in my steps/actions and have provided a minimal example. I kindly hope you can observe if either I'm doing something wrong or this is in fact a bug.

**Comment by EricDahlvang** [comment]
@tacri or @clearab any ideas?

**Comment by clearab** [comment]
I think the last step is the issue, that isn't uninstalling the bot from the team. That removes the app from your organizations app store, however it isn't uninstalled. Try:

1. Click on the ... overflow menu next to the team name
2. Pick Manage Team
3. Select the "Apps" tab
4. Click the trash icon to remove the app from that team
5. Then try to install the app again, your bot should receive the event

You could also try installing the app in a different team to see if the issue persists there as well.

**Comment by sferdi0** [comment]
Dear god... That worked.

1. I had several "test" apps registered under the Apps tab, installed during my various testing
2. This explains why I would sometimes get double notifications for Bot names that did not exist anymore
3. This also might explain why `on_teams_members_added` was not getting called correctly.

I cleaned the list up, uploaded the app;

```python3
async def on_teams_members_added(  # pylint: disable=unused-argument
    self,
    teams_members_added: [TeamsChannelAccount],
    team_info: TeamInfo,
    turn_context: TurnContext,
):
    for member in teams_members_added:
        if member.id == turn_context.activity.recipient.id and team_info is not None:
            # bot entered a Teams channel
            await turn_context.send_activity("Hello channel! I have just been added")
```
![](https://i.imgur.com/UQccdKL.png)

Thanks @rvinothrajendran @EricDahlvang @clearab

I will btw not eat my shoe, because one would assume that `Apps -> Delete` deletes the app - instead "uninstalling" is the correct method.

**Comment by rvinothrajendran** [comment]
@sferdi0 That’s great
Now I understand why it’s working for me always (AppStudio 😎)
Thanks @EricDahlvang @clearab

**Comment by clearab** [comment]
I completely agree - that remove from the app catalogue flow is really confusing. I've opened a bug on the team that owns the flow, and hopefully I can convince them to fix the wording to make it clearer what is actually happening there. Sorry about all the frustration.

**Comment by sferdi0** [comment]
Oh lovely, thanks!
